hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java

```diff
@@ -0,0 +1,849 @@
+
+            initPreparedStatements();
+
+            GsonBuilder gsonBuilder = new GsonBuilder();
+            gsonBuilder.registerTypeHierarchyAdapter(ConditionEval.class, new GsonAdapter<ConditionEval>());
+            gson = gsonBuilder.create();
+
+            reload();
+        } catch (Throwable t) {
+            msgLog.errorCannotInitializeAlertsService(t.getMessage());
+        }
+    }
+
+    private void initPreparedStatements() throws Exception {
+        if (insertAlert == null) {
+            insertAlert = session.prepare("INSERT INTO " + keyspace + ".alerts " +
+                    "(alertId, payload) VALUES (?, ?) ");
+        }
+        if (insertAlertTrigger == null) {
+            insertAlertTrigger = session.prepare("INSERT INTO " + keyspace + ".alerts_triggers " +
+                    "(alertId, triggerId) VALUES (?, ?) ");
+        }
+        if (insertAlertCtime == null) {
+            insertAlertCtime = session.prepare("INSERT INTO " + keyspace + ".alerts_ctimes " +
+                    "(alertId, ctime) VALUES (?, ?) ");
+
+        }
+        if (insertAlertStatus == null) {
+            insertAlertStatus = session.prepare("INSERT INTO " + keyspace + ".alerts_statuses " +
+                    "(alertId, status) VALUES (?, ?) ");
+        }
+        if (updateAlert == null) {
+            updateAlert = session.prepare("UPDATE " + keyspace + ".alerts " +
+                    "SET payload = ? WHERE alertId = ? ");
+        }
+        if (selectAlertStatus == null) {
+            selectAlertStatus = session.prepare("SELECT alertId, status FROM " + keyspace + ".alerts_statuses " +
+                    "WHERE alertId = ? ALLOW FILTERING ");
+        }
+        if (deleteAlertStatus == null) {
+            deleteAlertStatus = session.prepare("DELETE FROM " + keyspace + ".alerts_statuses " +
+                    "WHERE alertId = ? AND status = ?");
+        }
+    }
+
+    public void addAlerts(Collection<Alert> alerts) throws Exception {
+        if (alerts == null) {
+            throw new IllegalArgumentException("Alerts must be not null");
+        }
+        if (session == null) {
+            throw new RuntimeException("Cassandra session is null");
+        }
+        if (insertAlert == null) {
+            throw new RuntimeException("insertAlert PreparedStatement is null");
+        }
+        try {
+            for (Alert a : alerts) {
+                session.execute(insertAlert.bind(a.getAlertId(), toJson(a)));
+                session.execute(insertAlertTrigger.bind(a.getAlertId(), a.getTriggerId()));
+                session.execute(insertAlertCtime.bind(a.getAlertId(), a.getCtime()));
+                session.execute(insertAlertStatus.bind(a.getAlertId(), a.getStatus().name()));
+            }
+        } catch (Exception e) {
+            msgLog.errorDatabaseException(e.getMessage());
+            throw e;
+        }
+    }
+
+    public List<Alert> getAlerts(AlertsCriteria criteria) throws Exception {
```

Comment:
I see your point and I started to model in that way.
But there is a tradeoff, as I commented in one of Jay's comments, we have a requeriment of query information in the domain.
After investigate/read/learn, we have a tradeoff between the logic in the client, the design of the model in the persistence layer and the logic (queries) we send to the persistence layer. And everyone is strongly dependent of the context itself.
So, our main goal here is to move from a h2 database to a cassandra persistence, as first step.
Also, we are trying to put all information related to a same trigger into same partition, to avoid that we need to hit several partitions.
Also, the alerts domain is different of others in metrics/inventory as we are going to have a discrete number of definitions, and a very small fraction of alerts (comparing, for example with metrics).
So, I have identified some areas for improvement into the model, but also I want to define perf tests to get measures and to validate perf progress in overall (i.e. not only focus on cassandra, but also analyze the global logic, sometimes the improvements are not in the obvious place).
Also, we sill have query business requeriments that we want to maintain, so everything will help as an input in the improvement process.